README.md
# direct_menu
always have the app menu of owncloud on top

![Screenshot of direct_menu app](https://bitgrid.net/~jus/direct_menu.png)


## Installation

- Clone repository into the apps/ directory
- Enable app using owncloudconsole: ```sudo -u www-data php occ app:enable direct_menu```

## Contributors

- TheAgentK
- LukasReschke
- icewind1991
- JonesfromHan
- SonRiab
- eppfel


CHANGELOG
0.9.1
* Improve support for Nextcloud theming

0.9.0
* Support for Nextcloud

0.8.0
* Use default header-appname
* Change current app title style to match the default style
* App icon width and height re-added
* Add compatibility to owncloud 9.1 loading animations (and do not break backward compatiblity to 9.0!)
* Clicking a navigation item will show up the loading icon (works even in user menu)

0.7.1
* Fix title overlay style in owncloud 9

0.7.0
* Fix css for OwnCloud 9.0
* Move to semantic versioning
* Fixes scrolling bug on touch devices

0.6
* Fix icon sizing in Edge browser
* Change the look of the mouseover bubbles in menu 
* Move app label to the left of the icons

0.5
* Fix hidden menu toggle in owncloud version 8.2.1

0.4
* Specify OCS ID
* Changed menu text to speech bubbles

0.3
* Move appname to hover bubble
* Fix issue when using custom themes
* Add fallback for mobile devices

0.2
* Show app title when selected
* Show app title when hovering icon


appinfo/routes.php
<?php
/**
 * @copyright Copyright (c) 2016 Julius Härtl <jus@bitgrid.net>
 *
 * @author Julius Härtl <jus@bitgrid.net>
 *
 * @license GNU AGPL version 3 or any later version
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

return ['routes' => [
	[
		'name' => 'App#stylesheet',
		'url' => '/styles',
		'verb' => 'GET',
	]]
];

templates/direct_menu.php
@media only screen and (min-width: <?php p($_['width']); ?>px) {
	#navigation {
		margin: 0;
		padding: 0;
		position: absolute;
		top: 0;
		left: 80px;
		box-shadow: none;
		overflow: visible !important;
	}

	#navigation {
		width: auto;
		padding: 0;
		margin: 0;
		padding: 0;
		margin: 0;
		background-color: transparent;
		display: block !important;
		position: fixed;
	}

	#navigation #apps {
		overflow: visible;
	}

	#navigation div ul {
		position: relative;
		margin-left: 170px;
	}

	#navigation div ul:after {
		content: ".";
		clear: both;
		display: block;
		visibility: hidden;
		height: 0;
	}

	#navigation div ul li {
		float: left;
		padding: 0;
		margin: 0;
		height: 44px;
	}

	#navigation div ul li a {
		margin: 0;
		padding: 12px;
		padding-left: 6px;
		padding-right: 6px;
		width: 50px;
		height: 44px;
		position: static;
		text-align: center;
		color: #1d2d44 !important;
	}

	#navigation img.app-icon {
		width: 20px;
		height: 20px;
	}

	#navigation .app-loading .app-icon,
	#navigation .app-loading svg {
		display: none;
	}

	#navigation .app-loading .icon-loading-dark {
		background-size: 20px;
		position: relative;
		top: 0;
		left: 0;
		width: 20px;
		height: 20px;
		display: inline-block !important;
	}

	#navigation #apps .app-loading .icon-loading-dark::after {
		width: 18px;
		height: 18px;
		margin: -10px 0 0 -10px;
	}

	#navigation div li span {
		display: none;
		position: absolute;
		overflow: visible;
		left: 0;
		top: 45px;
		background-color: #FFF;
		border: 1px solid #000;
		color: #000;
		min-width: 120px;
	}

	#navigation div li:hover a {
		position: relative
	}

	#navigation div li:hover span {
		display: block
	}

	#navigation div li:hover span {
		background-color: white;
		-webkit-border-radius: 3px;
		-moz-border-radius: 3px;
		border-radius: 3px;
		margin-top: 3px;
		border: none;
		color: #333;
		left: 50%;
		margin-left: -60px;
		padding: 5px;
		-webkit-filter: drop-shadow(0 0 5px rgba(150, 150, 150, 0.75));
		-moz-filter: drop-shadow(0 0 5px rgba(150, 150, 150, 0.75));
		-ms-filter: drop-shadow(0 0 5px rgba(150, 150, 150, 0.75));
		-o-filter: drop-shadow(0 0 5px rgba(150, 150, 150, 0.75));
		filter: drop-shadow(0 0 5px rgba(150, 150, 150, 0.75));
	}

	#navigation div li:hover span {
		display: block;
	}

	#navigation div li:hover span:after,
	#navigation div li:hover span:before {
		content: " ";
		bottom: 100%;
		left: 50%;
		border: solid transparent;
		height: 0;
		width: 0;
		position: absolute;
		pointer-events: none;
		margin-left: -9px;
	}

	#navigation div li:hover span:after {
		border-bottom-color: white;
		border-width: 8px;
		margin-left: -8px;
	}

	#navigation div li:hover span:before {
		border-bottom-color: white;
		border-width: 9px;
	}

	a.menutoggle .icon-caret {
		display: none !important;
	}

	#navigation svg {
		width: 20px;
		height: 20px;
	}

	#navigation svg image {
		width: 100%;
		height: 100%;
<?php if($_['inverted']===false) { ?>
		-webkit-filter: none;
		filter: none;
<?php } ?>
	}

	#navigation a svg,
	#navigation a span {
		-ms-filter: "progid:DXImageTransform.Microsoft.Alpha(Opacity=90)";
		opacity: .9;
	}

	#navigation a:hover svg,
	#navigation a:hover span,
	#navigation a.active svg,
	#navigation a.active span {
		-ms-filter: "progid:DXImageTransform.Microsoft.Alpha(Opacity=100)";
		opacity: 1;
	}
}


controller/appcontroller.php
<?php
/**
 * @copyright Copyright (c) 2016 Julius Härtl <jus@bitgrid.net>
 *
 * @author Julius Härtl <jus@bitgrid.net>
 *
 * @license GNU AGPL version 3 or any later version
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

namespace OCA\Direct_menu\Controller;
use OCP\AppFramework\Http\TemplateResponse;
use OCP\AppFramework\Http\DataDownloadResponse;
use OCP\AppFramework\Utility\ITimeFactory;
use OCP\IRequest;

class AppController extends \OCP\AppFramework\Controller {

	public function __construct($appName, IRequest $request, ITimeFactory $timeFactory) {
		parent::__construct($appName, $request);
		$this->timeFactory = $timeFactory;
	}

	/**
	 * @NoCSRFRequired
	 * @PublicPage
	 *
	 * @return DataDownloadResponse
	 */
	public function stylesheet() {
		$inverted = false;
		if(\OCP\App::isEnabled('theming') && class_exists('\OCA\Theming\Util')) {
			$color = \OC::$server->getThemingDefaults()->getMailHeaderColor();
			$util = new \OCA\Theming\Util();
			$inverted = $util->invertTextColor($color);
		}

		$navigation = \OC::$server->getNavigationManager()->getAll();
		$navigationCount = count($navigation);

		// 250px for icon/appname
		// 120px for user menu + 1 icon spacing
		$width = $navigationCount*50+250+170;

		$params = [
			'width' => $width,
			'inverted' => $inverted
		];
		$template = new TemplateResponse('direct_menu', 'direct_menu', $params, 'blank');
		$response = new DataDownloadResponse($template->render(), 'style', 'text/css');
		$response->addHeader('Expires', date(\DateTime::RFC2822, $this->timeFactory->getTime()));
		$response->addHeader('Pragma', 'cache');
		$response->cacheFor(3600);
		return $response;
	}

}

lib/Controller/AppController.php
<?php
/**
 * @copyright Copyright (c) 2016 Julius Härtl <jus@bitgrid.net>
 *
 * @author Julius Härtl <jus@bitgrid.net>
 *
 * @license GNU AGPL version 3 or any later version
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Affero General Public License as
 *  published by the Free Software Foundation, either version 3 of the
 *  License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Affero General Public License for more details.
 *
 *  You should have received a copy of the GNU Affero General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

namespace OCA\Direct_menu\Controller;
use OCP\AppFramework\Http\DataDisplayResponse;
use OCP\AppFramework\Http\TemplateResponse;
use OCP\AppFramework\Http\DataDownloadResponse;
use OCP\AppFramework\Utility\ITimeFactory;
use OCP\IRequest;

class AppController extends \OCP\AppFramework\Controller {

	public function __construct($appName, IRequest $request, ITimeFactory $timeFactory) {
		parent::__construct($appName, $request);
		$this->timeFactory = $timeFactory;
	}

	/**
	 * @NoCSRFRequired
	 * @PublicPage
	 *
	 * @return DataDownloadResponse
	 */
	public function stylesheet() {
		$inverted = false;
		if(\OCP\App::isEnabled('theming')) {
			$color = \OC::$server->getThemingDefaults()->getMailHeaderColor();
			$util = new \OCA\Theming\Util();
			$inverted = $util->invertTextColor($color);
		}

		$navigation = \OC::$server->getNavigationManager()->getAll();
		$navigationCount = count($navigation);

		// 250px for icon/appname
		// 120px for user menu + 1 icon spacing
		$width = $navigationCount*50+250+170;

		$params = [
			'width' => $width,
			'inverted' => $inverted
		];
		$template = new TemplateResponse('direct_menu', 'direct_menu', $params, 'blank');
		$response = new DataDownloadResponse($template->render(), 'style', 'text/css');
		$response->addHeader('Expires', date(\DateTime::RFC2822, $this->timeFactory->getTime()));
		$response->addHeader('Pragma', 'cache');
		$response->cacheFor(3600);
		return $response;
	}

}